general/mem_manager.hpp

```diff
@@ -78,8 +78,10 @@ class mm
 
    /// Translates ptr to host or device address, depending on
    /// Device::UsingDevice() and the ptr state.
-   static inline void *ptr(void *a) { return MM().Ptr(a); }
-   static inline const void* ptr(const void *a) { return MM().Ptr(a); }
+   template <class T>
+   static inline T *ptr(T *a) { return static_cast<T*>(MM().Ptr(a)); }
+   template <class T>
+   static inline const T* ptr(const T *a) { return static_cast<const T*>(MM().Ptr(a)); }
```

Comment:
You probably do not need a special `const` version -- the general version should work fine with `T = const U`.